binary-oracle-pair/program/src/instruction.rs

```diff
@@ -0,0 +1,211 @@
+pub fn deposit(
+    program_id: &Pubkey,
+    pool: &Pubkey,
+    authority: &Pubkey,
+    user_transfer_authority: &Pubkey,
+    is_user_authority_signer: bool,
+    user_token_account: &Pubkey,
+    pool_deposit_token_account: &Pubkey,
+    token_pass_mint: &Pubkey,
+    token_fail_mint: &Pubkey,
+    token_pass_destination_account: &Pubkey,
+    token_fail_destination_account: &Pubkey,
+    token_program_id: &Pubkey,
+    amount: u64,
+) -> Result<Instruction, ProgramError> {
+    let init_data = PoolInstruction::Deposit(amount);
+    let data = init_data.try_to_vec()?;
+    let accounts = vec![
+        AccountMeta::new_readonly(*pool, false),
+        AccountMeta::new_readonly(*authority, false),
+        AccountMeta::new_readonly(*user_transfer_authority, is_user_authority_signer),
+        AccountMeta::new(*user_token_account, false),
+        AccountMeta::new(*pool_deposit_token_account, false),
+        AccountMeta::new(*token_pass_mint, false),
+        AccountMeta::new(*token_fail_mint, false),
+        AccountMeta::new(*token_pass_destination_account, false),
+        AccountMeta::new(*token_fail_destination_account, false),
+        AccountMeta::new_readonly(sysvar::clock::id(), false),
+        AccountMeta::new_readonly(*token_program_id, false),
+    ];
+    Ok(Instruction {
+        program_id: *program_id,
+        accounts,
+        data,
+    })
+}
+
+/// Create `Withdraw` instruction
+#[allow(clippy::too_many_arguments)]
+pub fn withdraw(
+    program_id: &Pubkey,
+    pool: &Pubkey,
+    authority: &Pubkey,
+    pool_deposit_token_account: &Pubkey,
+    token_pass_user_account: &Pubkey,
+    token_fail_user_account: &Pubkey,
+    token_pass_mint: &Pubkey,
+    token_fail_mint: &Pubkey,
+    user_token_destination_account: &Pubkey,
+    token_program_id: &Pubkey,
+    amount: u64,
+) -> Result<Instruction, ProgramError> {
+    let init_data = PoolInstruction::Withdraw(amount);
+    let data = init_data.try_to_vec()?;
+    let accounts = vec![
+        AccountMeta::new_readonly(*pool, false),
+        AccountMeta::new_readonly(*authority, false),
+        AccountMeta::new(*pool_deposit_token_account, false),
+        AccountMeta::new(*token_pass_user_account, false),
+        AccountMeta::new(*token_fail_user_account, false),
+        AccountMeta::new(*token_pass_mint, false),
+        AccountMeta::new(*token_fail_mint, false),
+        AccountMeta::new(*user_token_destination_account, false),
+        AccountMeta::new_readonly(sysvar::clock::id(), false),
+        AccountMeta::new_readonly(*token_program_id, false),
+    ];
+    Ok(Instruction {
+        program_id: *program_id,
+        accounts,
+        data,
+    })
+}
+
+/// Create `Decide` instruction
+pub fn decide(
+    program_id: &Pubkey,
+    pool: &Pubkey,
+    decider: &Pubkey,
+    decision: bool,
+) -> Result<Instruction, ProgramError> {
+    let init_data = PoolInstruction::Decide(decision);
+    let data = init_data.try_to_vec()?;
```

Comment:
your know the size. known size struct(it may be even on stack). unwrap borsch serde into it. there seems not reason to allocated arbitraty heap here (and hence no error code).